### ISSUE TITLE
Doom Emacs is unbelievably slow

### ISSUE THREAD

**Issue by egordiac** [hlissner/doom-emacs#4320]
**What did you expect to happen?**
I expected to open Doom Emacs, create/open a new file and start editing it, while the editor remains fairly snappy and fast.

**What actually happened?**
I tried editing the default init.el, just to test out Doom, but after pressing enter to open the file the editor froze for around 10 seconds. Then, when the editor unfroze, I realized the typing is very slow, not only that but the hjkl keys and M-x commands all type in very slowly. Also when scrolling or using the mouse there's lots of input lag as well.

**Additional details:**
I've tried using `so-long-mode` and `so-long-minor-mode,` but to no avail, and I would prefer to refrain from turning off the autocomplete, so I don't want to use `so-long-mode`.
`view-echo-area-messages` outputted this: 'Doom loaded 159 packages across 29 modules in 1.449s', which shows that the editor starts up and loads everything pretty quickly, so I don't know why the rest of the editor isn't like this.

**Steps to reproduce:**
1. Install Emacs (version 27.1-3) in my case.
2. Install Doom.
3. Start Doom and try editing a file.

**System information:**
<details><pre>
```
SYSTEM  type       gnu/linux
        config     x86_64-pc-linux-gnu
        shell      /bin/bash
        uname      Linux 5.4.78-1-MANJARO #1 SMP PREEMPT Wed Nov 18 21:31:46 UTC 2020 x86_64
        path       (~/.emacs.d/bin/ ~/.local/bin /usr/local/sbin /usr/local/bin /usr/bin /usr/lib/jvm/default/bin /usr/bin/site_perl /usr/bin/vendor_perl /usr/bin/core_perl /var/lib/snapd/snap/bin /usr/lib/emacs/27.1/x86_64-pc-linux-gnu/)
EMACS   dir        ~/.emacs.d/
        version    27.1
        build      Aug 28, 2020
        buildopts  --prefix=/usr --sysconfdir=/etc --libexecdir=/usr/lib --localstatedir=/var --with-x-toolkit=gtk3 --with-xft --with-wide-int --with-modules --with-cairo --with-harfbuzz 'CFLAGS=-march=x86-64 -mtune=generic -O2 -pipe -fno-plt' CPPFLAGS=-D_FORTIFY_SOURCE=2 LDFLAGS=-Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now
        features   XPM JPEG TIFF GIF PNG RSVG CAIRO SOUND GPM DBUS GSETTINGS GLIB NOTIFY INOTIFY ACL GNUTLS LIBXML2 FREETYPE HARFBUZZ M17N_FLT LIBOTF ZLIB TOOLKIT_SCROLL_BARS GTK3 X11 XDBE XIM MODULES THREADS LIBSYSTEMD JSON PDUMPER LCMS2 GMP
        traits     (gui server-running envvar-file)
DOOM    dir        ~/.doom.d/
        version    2.0.9
        font       #<font-object -ADBO-Source Code Pro-normal-normal-normal-*-13-*-*-*-m-0-iso10646-1>
        theme      doom-one
        build      grafted, HEAD -> develop, origin/develop, origin/HEAD e4890f6 2020-11-25 00:40:43 -0500
        elc-files  0
        modules    (:completion company ivy :ui doom doom-dashboard doom-quit hl-todo modeline ophints (popup +defaults) vc-gutter vi-tilde-fringe workspaces :editor (evil +everywhere) file-templates fold snippets :emacs dired electric undo vc :checkers syntax :tools (eval +overlay) lookup magit :lang emacs-lisp markdown org sh :config (default +bindings +smartparens))
        packages   (n/a)
        unpin      (n/a)
        elpa       (n/a)
```
</pre></details>

**Comment by hlissner** [comment]
I cannot reproduce this on NixOS or any of my linux VMs (even on my slowest VM I'm seeing a pause of 3-4s). It's expected that the first file you load will take a while (1-2 seconds on faster hardware, 2-4s on slower), but 10s is beyond normal. 

Can you tell me about the hardware you're running this on, or the environment it's running in? Perhaps screencast this?

Here are some things you should suspect:

1. A rogue compositor preventing Emacs from redrawing as often as it should. Try disabling compton/picom if you have it running.
2. If you're not on SSDs or on very old harddrives, they could be the bottleneck.
3. Anti-virus/malware? This can easily add seconds to Emacs file IO. Whitelist `~/.emacs.d`, `~/.doom.d`, and site files if possible.
4. If your gpu has vsync on, try adding `(add-to-list 'default-frame-alist '(inhibit-double-buffering . t))` to `~/.doom.d/config.el`
5. If you don't have gpu drivers installed, install them. I've heard performance complaints from folks with hiDPI displays without gpu drivers installed.

If none of those help, I'm afraid I'm out of ideas.

**Comment by egordiac** [comment]
Running Manjaro linux with an AMD A8 and Radeon R5 and open source drivers.

1. I don't think I even have picom or compton installed, so that shouldn't be a problem.
2. This laptop has a hard drive, so that could be why.
3. I don't have any antimalware programs installed on here.
4. I've tried, but that doesn't seem to do anything.
5. I've tried open source and proprietary drivers.
6. Tried installing Emacs 26.3, but the issue hasn't changed.

I don't know what could be causing this, because normal Emacs installations are very snappy and fast with seemingly no input lag. Maybe it's something else, but for now I'll try using the build of Emacs you suggested, however that isn't likely to help.

**Comment by tecosaur** [comment]
Have you tried profiling Emacs when it's being slow? `SPC h T` to start and end the profile, `TAB` to expand sections till you find what's eating the performance.

**Comment by hexop** [comment]
I've had some freezes with ligatures from time to time. When that happens emacs jumps to 100% CPU usage and it doesn't respond, even while using `C-g`.
I'm using emacs-pgtk-native-comp, but it happened to me on normal native-comp and on emacs 27.
The only workaround I've found is to disable them.

**Comment by linges** [comment]
I have the same problem when I open the `init.el` with a complete fresh doom setup. 
I'm running arch linux. I will try to start the profiler but usually I have to kill emacs because not even `C-g` helps.
I have been using emacs for a long time without issues, so I assume its not my hardware.

**Comment by linges** [comment]
During profiler run I once saw this:
`find-function-library: Symbol’s function definition is void: redisplay_internal\ \(C\ function\)`
The output:
[profiler.txt](https://github.com/hlissner/doom-emacs/files/5630797/profiler.txt)

**Comment by linges** [comment]
Short update: I saw similar behavior with plain emacs, so maybe not doom related.

Additionally it does not happen with an emacs that uses lucid instead of gtk.

**Comment by wnkw** [comment]
I believe this is similar to the issue I created [here](https://github.com/syl20bnr/spacemacs/issues/14462). This is only on emacs 27 using spacemacs but it might be related to doom-theme I was using.

**Comment by lebensterben** [comment]
@wnkw I use doom-one-dark theme in spacemacs and I have no issue at all.

**Comment by wnkw** [comment]
@lebensterben  Yes, sorry this was a bit rushed. Disabling doom themes and mode-line didn't help in my case either.

**Comment by linges** [comment]
I tracked my issue down to unicode symbols and missing fonts. The symbol in question was a :smile: 
This stackexhange helped me figuring it out: https://emacs.stackexchange.com/questions/61215/paste-very-specific-unicode-in-scratch-buffer-freezes-emacs-q 
In the end I installed the fonts recommaned by https://github.com/rolandwalker/unicode-fonts and that was already enough to solve it.
